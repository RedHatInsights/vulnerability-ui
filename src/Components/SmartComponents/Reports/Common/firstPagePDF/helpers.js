import messages from '../../../../../Messages';

export const getAffectingFilterTexts = (systemCount) => ({
    false: `not affecting any ${systemCount}`,
    true: `impacting at least one of your ${systemCount} analyzed RHEL systems.`,
    rpmdnf: `impacting at least one of your analyzed ${systemCount} RHEL conventional systems.`,
    edge: 'impacting at least one of your analyzed ${systemCount} RHEL immutable systems.',
    'rpmdnf,edge': `impacting at least one of your analyzed ${systemCount} RHEL systems`
});

export const cveFilterStringSegments = [
    { name: 'impact', msg: messages.CVEsPDFreportFiltersSeverity, lowercase: true },
    { name: 'business_risk_id', msg: messages.CVEsPDFreportFiltersBusinessRisk, lowercase: true },
    { name: 'status_id', msg: messages.CVEsPDFreportFiltersStatus, lowercase: true },
    { name: 'rule_presence', msg: messages.CVEsPDFreportFiltersDefinedValue, lowercase: true },
    { name: 'known_exploit', msg: messages.CVEsPDFreportFiltersDefinedValue, lowercase: true },
    { name: 'advisory_available', msg: messages.CVEsPDFreportFiltersAdvisoryAvailable, lowercase: true },
    { name: 'rhel_version', msg: messages.CVEsPDFreportFiltersOsVersion }
];

export const systemsFilterStringSegments = [
    { name: 'filter', msg: messages.systemsPDFreportFiltersSearchTerm, lowercase: true },
    { name: 'rhel_version', msg: messages.systemsPDFreportFiltersOsVersion }
];
