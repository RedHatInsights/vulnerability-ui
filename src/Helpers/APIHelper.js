let BASE_ROUTE = '/api/vulnerability/v1';
import { GitApi } from '@redhat-cloud-services/vulnerabilities-client';
import { SystemProfileApi } from '@redhat-cloud-services/host-inventory-client';
import instance from '../Utilities/interceptors';
import { constructParameters } from './MiscHelper';

let api = new GitApi(undefined, BASE_ROUTE, instance);

const INVENTORY_API_BASE = '/api/inventory/v1';
const systemProfile = new SystemProfileApi(undefined, INVENTORY_API_BASE, instance);

const defaultParams = [
    'filter',
    'limit',
    'offset',
    'page',
    'page_size',
    'sort'
];

export function getCveDetails(synopsis) {
    const result = api.getCveDetails(synopsis);
    return result;
}

export function getSystemDetails(system) {
    const result = api.getSystemDetails(system);
    return result;
}

export function setSystemCveStatus(params) {
    let result = api.setStatus(params);
    return result;
}

export function optOutSystems(systems, optOut) {
    const data = { opt_out: optOut, inventory_id: new Array(...systems) };
    let result = api.setSystemsOptOut(data);
    return result;
}

export function setBusinessRisk(params) {
    let result = api.setCveBusinessRisk(params);
    return result;
}

export function setCveStatus(params) {
    let result = api.setCveStatus(params);
    return result;
}

/**
 * /vulnerabilities/cves
 * /vulnerabilities/cves/ids
 */
let cveParams = [
    'cvss_from',
    'cvss_to',
    'public_from',
    'public_to',
    'impact',
    'data_format',
    'business_risk_id',
    'status_id',
    'rule_presence',
    'tags',
    'sap_sids',
    'sap_system',
    'known_exploit',
    'affecting',
    'rhel_version'
];

export function getCveListByAccount(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...cveParams,
        'report',
        'advanced_report',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getCveList(...parameterArray);
    return result;
}

export function getCveIdsList(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...cveParams,
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getCveIdsList(...parameterArray);
    return result;
}

/**
 * /cves/ID/affected_systems
 */
let systemsByCVEparams = [
    'status_id',
    'data_format',
    'uuid',
    'rule_key',
    'rule_presence',
    'rule',
    'tags',
    'sap_sids',
    'sap_system',
    'show_advisories',
    'advisory',
    'rhel_version',
    'first_reported_from',
    'first_reported_to',
    'advisory_available',
    'remediation'
];

export function getAffectedSystemsByCVE({ id, ...apiProps }) {
    let parameterNames = [
        ...defaultParams,
        ...systemsByCVEparams,
        'report',
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getAffectedSystemsByCve(id, ...parameterArray);
    return result;
}

export function getAffectedSystemsIdsByCve({ id, ...apiProps }) {
    let parameterNames = [
        ...defaultParams,
        ...systemsByCVEparams,
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getAffectedSystemsIdsByCve(id, ...parameterArray);
    return result;
}

/**
 * /systems/ID/cves
 * /systems/ID/cves/ids
 */
let cveBySystemParams = [
    'cvss_from',
    'cvss_to',
    'public_from',
    'public_to',
    'impact',
    'status_id',
    'data_format',
    'business_risk_id',
    'rule_presence',
    'show_advisories',
    'advisory',
    'rule_key',
    'known_exploit',
    'first_reported_from',
    'first_reported_to',
    'advisory_available',
    'remediation'
];

export function getCveListBySystem(apiProps) {
    const { system } = apiProps;
    let parameterNames = [
        ...defaultParams,
        ...cveBySystemParams,
        'report'
    ];

    if (apiProps && system) {
        Object.keys(apiProps).forEach(key => (apiProps[key] === undefined || apiProps[key] === '') && delete apiProps[key]);
        const params = parameterNames.map(item => apiProps[item]);
        return api.getCveListBySystem(system, ...params).catch(err => {
            if (err && err.status === '404') {
                return { errors: err };
            }

            throw err;
        });
    }
}

export function getCveIdsBySystem(apiProps) {
    const { system } = apiProps;
    let parameterNames = [
        ...defaultParams,
        ...cveBySystemParams
    ];

    if (apiProps && system) {
        Object.keys(apiProps).forEach(key => (apiProps[key] === undefined || apiProps[key] === '') && delete apiProps[key]);
        const params = parameterNames.map(item => apiProps[item]);
        return api.getCveIdsBySystem(system, ...params).catch(err => {
            if (err && err.status === '404') {
                return { errors: err };
            }

            throw err;
        });
    }
}

/**
 * /systems/ids
 * /systems
 */
let systemsParams = [
    'data_format',
    'stale',
    'uuid',
    'tags',
    'sap_sids',
    'sap_system',
    'excluded',
    'rhel_version'
];

export function getSystems(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...systemsParams,
        'report',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getSystemsList(...parameterArray);
    return result;
}

export function getSystemsIds(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...systemsParams
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getSystemsIds(...parameterArray);
    return result;
}

export function getAnnouncement() {
    let result = api.getAnnouncement();
    return result;
}

export function getDashbar(apiProps) {
    let parameterNames = [
        'tags',
        'sap_sids',
        'sap_system',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getDashbar(...parameterArray);
    return result;
}

export function getOperatingSystems() {
    return systemProfile.apiSystemProfileGetOperatingSystem();
}

export function fetchPdfReport(template, filters) {
    const CRC_PDF_GENERATE_API = '/api/crc-pdf-generator/v1/generate';
    const url = new URL(CRC_PDF_GENERATE_API,  window.location.origin);

    return fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },

        body: JSON.stringify({
            service: 'vulnerability',
            template,
            params: {
                ...filters
            }
        })
    }, 50000)
        .then((response) => response.blob());
}
