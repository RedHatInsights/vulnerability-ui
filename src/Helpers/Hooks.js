import union from 'lodash/union';
import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { useIntl } from 'react-intl';
import {
    addNotification as addNotificationFEC,
    clearNotifications as clearNotificationsFEC
} from '@redhat-cloud-services/frontend-components-notifications/redux';
import { optOutSystemsAction } from '../../src/Store/Actions/Actions';
import messages from '../Messages';

export const useNotification = (config) => {
    const dispatch = useDispatch();

    const globalConfig = {
        variant: 'info',
        ...config
    };

    useEffect(() => {
        return () => {
            if (!config.preventClear) {
                dispatch(clearNotificationsFEC());
            }
        };
    }, [dispatch]);

    const addNotification = ({ msg, ...config }) =>
        dispatch(
            addNotificationFEC({
                title: msg,
                ...globalConfig,
                ...config
            })
        );

    const clearNotifications = () =>
        dispatch(clearNotificationsFEC());

    return [addNotification, clearNotifications];
};

export const useGetEntities = (fetchApi, setUrlParams) => {

    const getEntities = async (
        _items,
        { orderBy, orderDirection, page, per_page: perPage, vulnerabilityParams }
    ) => {
        const sort = `${orderDirection === 'ASC' ? '' : '-'}${orderBy}`;

        const params = {
            ...vulnerabilityParams,
            page,
            page_size: perPage,
            sort
        };

        setUrlParams({ ...params });

        const items = await fetchApi({ ...params });

        return {
            results: items?.data?.map(row => ({ id: row.id, ...row.attributes })),
            total: items?.meta?.total_items
        };
    };

    return getEntities;
};

export const useOptOutSystems = onRefreshInventory => {
    const [addSuccessNotification] = useNotification({ variant: 'success' });
    const [addFailureNotification] = useNotification({ variant: 'danger', autoDismiss: false });
    const dispatch = useDispatch();
    const intl = useIntl();

    /**
     * Excludes/included provided system(s), if toggling one system systemId and systemName have to be set.
     * If toggling multiple systems, these two params are ignored and system info will be extracted from selectedRows.
     * @param {object} affectedRows - object with keys representing ids of systems to opt out/in
     * @param {?string} systemName - display name of a system if only a single system is toggled
     * @param {bool} isIncluded - if true systems will be excluded, else included
     */
    return (affectedRows, systemName, isIncluded) => {
        const count = Object.keys(affectedRows).length;

        dispatch(optOutSystemsAction(Object.keys(affectedRows), isIncluded))
            .then(() => {
                isIncluded ?
                    addSuccessNotification({
                        msg: intl.formatMessage(messages.notificationExcludeSuccess, { count, systemName })
                    })
                    : addSuccessNotification({
                        msg: intl.formatMessage(messages.notificationIncludeSuccessTitle, { count, systemName }),
                        description: intl.formatMessage(messages.notificationIncludeSuccessBody)
                    });
            }).catch(() => {
                isIncluded ?
                    addFailureNotification({
                        msg: intl.formatMessage(messages.notificationExcludeFailureTitle, { count, systemName }),
                        description: intl.formatMessage(messages.notificationExcludeFailureBody, { count })
                    })
                    : addFailureNotification({
                        msg: intl.formatMessage(messages.notificationIncludeFailureTitle, { count, systemName }),
                        description: intl.formatMessage(messages.notificationIncludeFailureBody, { count })
                    });
            }).finally(onRefreshInventory);
    };
};

const mapSelectedRows = ({ id }) => ({ id, selected: true });

export const useBulkSelect = ({ rawData, selectedRows, selectedRowsCount, handleSelect, fetchResource, multiRow = false }) => {
    const intl = useIntl();
    const { meta, data } = rawData;

    const selectPageNumber = data && (((Array.isArray(data) && data.length) && multiRow) ? data.length / 2 : data.length);

    const handleSelectPage = () => {
        const { data } = rawData ?? {};
        const collectedRows = data && union(
            data.filter(item => item.id).map(mapSelectedRows),
            Object.values(selectedRows)
        );

        handleSelect(collectedRows);
    };

    const handleUnselect = () => handleSelect([]);

    const handleSelectAll = async () => {

        let { payload } = await fetchResource({ page_size: meta.total_items, page: 1 });

        payload.then(({ data }) => {
            handleSelect(data.map(mapSelectedRows));
        });
    };

    return {
        count: selectedRowsCount,
        checked: Boolean(selectedRowsCount),
        isDisabled: meta.total_items === 0 && selectedRowsCount === 0,
        ouiaId: 'bulk-select',
        items: [
            {
                title: intl.formatMessage(messages.selectNone),
                onClick: handleUnselect
            },
            {
                title: `${intl.formatMessage(messages.selectPage, { count: selectPageNumber })}`,
                onClick: handleSelectPage
            },
            {
                title: `${intl.formatMessage(messages.selectAll, { count: meta.total_items ? meta.total_items : 0 })}`,
                onClick: handleSelectAll
            }
        ]

    };

};
