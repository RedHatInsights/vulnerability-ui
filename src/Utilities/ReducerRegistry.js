import promiseMiddleware from 'redux-promise-middleware';
import { getRegistry } from '@red-hat-insights/insights-frontend-components';

const registry = getRegistry({}, [promiseMiddleware()]);

export default {
    getStore: () => registry.getStore(),
    register: (newReducers) => {
        registry.register(newReducers);
    }
};
