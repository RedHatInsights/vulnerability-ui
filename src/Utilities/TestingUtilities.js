import React from 'react';
import PropTypes from 'prop-types';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { IntlProvider } from 'react-intl';

export const ComponentWithContext = ({
    Component,
    componentProps,
    renderOptions = {}
}) => {
    const mockStore = configureStore();

    return (
        <IntlProvider locale="en">
            <Provider store={renderOptions?.store || mockStore()}>
                <MemoryRouter initialEntries={renderOptions?.initialEntries || ['/']}>
                    {renderOptions?.componentPath ? (
                        <Routes>
                            <Route>
                                <Component {...componentProps} />
                            </Route>
                        </Routes>
                    ) : (
                        <Component {...componentProps} />
                    )}
                </MemoryRouter>
            </Provider>
        </IntlProvider>
    );
};

ComponentWithContext.propTypes = {
    Component: PropTypes.element,
    componentProps: PropTypes.object,
    renderOptions: PropTypes.object
};
