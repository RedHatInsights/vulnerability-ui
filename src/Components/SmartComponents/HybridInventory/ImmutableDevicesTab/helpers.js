import { getAffectedSystemsByCVE, useGetImageData } from '../../../../Helpers/APIHelper';
import { SYSTEMS_EXPOSED_HEADER } from '../../../../Helpers/constants';

export const fetchFullDeviceInfo = async (parameters, fetchImagesData) => {
    const vulnerableSystems = await getAffectedSystemsByCVE(parameters);
    const deviceIds = vulnerableSystems?.data.map(data => data.id);

    if (deviceIds.length) {
        const deviceData = await fetchImagesData({ devices_uuid: deviceIds });
        vulnerableSystems.data?.forEach(system => {
            const systemDevice = deviceData?.data?.devices.find(device => device.DeviceUUID === system.id);
            return system.attributes = { ...system.attributes, ...systemDevice };
        });
    }

    return vulnerableSystems;
};

export const useGetEntities = ({ id, setUrlParams, createRows }) => {
    const fetchImagesData = useGetImageData();
    const getEntities = async (
        _items,
        { orderBy, orderDirection, page, per_page: perPage, edgeParams, filters }
    ) => {
        const { hostnameOrId: filter } = filters;
        const sort = `${orderDirection === 'ASC' ? '' : '-'}${orderBy}`;

        const params = {
            ...edgeParams,
            page,
            page_size: perPage,
            sort,
            filter,
            ...filters?.hostGroupFilter ? {
                group_names: filters.hostGroupFilter.join(',')
            } : {}
        };

        setUrlParams?.({ ...params });

        const items = await fetchFullDeviceInfo(
            {
                ...id && { id },
                ...params
            },
            fetchImagesData
        );

        return {
            results: typeof createRows === 'function'
                ? createRows(items || {})
                : items?.data?.map(row => ({ id: row.id, ...row.attributes })),
            total: items?.meta?.total_items,
            page: items?.meta?.page,
            perPage: items?.meta?.page_size,
            meta: {
                page: items?.meta?.page,
                sort: items?.meta?.sort,
                page_size: items?.meta?.page_size,
                cves_without_errata: items?.meta?.cves_without_errata
            }
        };
    };

    return getEntities;
};

export const mergeAppColumns = (defaultColumns) => {
    const osColumn = SYSTEMS_EXPOSED_HEADER.find((column) => column.key === 'os');
    const advisoryColumn = SYSTEMS_EXPOSED_HEADER.find((column) => column.key === 'advisories_list');

    const osIndex = defaultColumns.findIndex(column =>{
        return column.key === 'system_profile';
    });
    if (osIndex !== -1) {
        defaultColumns[osIndex] = osColumn;
    }

    return [...defaultColumns, ...[advisoryColumn || []]];
};
