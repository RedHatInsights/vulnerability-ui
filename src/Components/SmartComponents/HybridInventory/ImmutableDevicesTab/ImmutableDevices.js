import React, { useEffect } from 'react';
import propTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { useUrlParams } from '../../../../Helpers/MiscHelper';
import { useDispatch, useSelector } from 'react-redux';
import useInsightsNavigate from '@redhat-cloud-services/frontend-components-utilities/useInsightsNavigate';
import ReducerRegistry from '../../../../Utilities/ReducerRegistry';
import { createExposedDevicesRows } from '../../../../Helpers/CVEHelper';
import { inventoryEntitiesReducer } from '../../../../Store/Reducers/InventoryEntitiesReducer';
import {
    changeExposedDevicesParameters
} from '../../../../Store/Actions/Actions';
import {
    SYSTEMS_EXPOSED_ALLOWED_PARAMS,
    PERMISSIONS,
    DEFAULT_PAGE_SIZE,
    RULE_PRESENCE_OPTIONS,
    CVE_DETAILS_FILTER_PARAMS
} from '../../../../Helpers/constants';
import ErrorHandler from '../../../PresentationalComponents/ErrorHandler/ErrorHandler';
import { EmptyStateNoSystems } from '../../../PresentationalComponents/EmptyStates/EmptyStates';
import { useGetEntities, mergeAppColumns } from './helpers.js';
import Spinner from '@redhat-cloud-services/frontend-components/Spinner';
import { buildActiveFilters, removeFilters, isFilterInDefaultState } from '../../../../Helpers/TableToolbarHelper';
import useSearchFilter from '../../../PresentationalComponents/Filters/PrimaryToolbarFilters/SearchFilter';
import useSecurityRuleFilter from '../../../PresentationalComponents/Filters/PrimaryToolbarFilters/SecurityRuleFilter';
import AsynComponent from '@redhat-cloud-services/frontend-components/AsyncComponent';
import { useRbac } from '../../../../Helpers/Hooks';
import messages from '../../../../Messages';

const ImmutableDevices = ({ intl, cveName, filterRuleValues, inventoryRef, headerFilters }) => {
    const [[canReadHostsInventory], isLoadingInventory] = useRbac([
        PERMISSIONS.readHosts
    ], 'inventory');

    const dispatch = useDispatch();
    const navigate = useInsightsNavigate('inventory');
    const [urlParameters, setUrlParams] = useUrlParams(SYSTEMS_EXPOSED_ALLOWED_PARAMS);

    const totalItems = useSelector(({ entities }) => entities?.total);
    const error = useSelector(({ entities }) => entities?.error || {});

    const parameters = useSelector(
        ({ ImmutableDevicesStore }) => ImmutableDevicesStore.parameters);

    const apply = (params) => dispatch(changeExposedDevicesParameters(params));

    useEffect(() => apply(urlParameters), []);

    useEffect(() => setUrlParams({ ...parameters, ...headerFilters }), [parameters, headerFilters]);

    const getEntities = useGetEntities(
        {
            id: cveName,
            createRows: createExposedDevicesRows,
            apply
        }
    );

    useEffect(() => {
        apply(headerFilters);
    }, [headerFilters]);

    const searchFilter = useSearchFilter(
        'filter',
        messages.systemsSearchName,
        messages.searchFilterByName,
        parameters.filter,
        apply
    );

    const advisoryNameFilter = useSearchFilter(
        'advisory',
        messages.advisoryName,
        messages.searchFilterByAdvisoryName,
        parameters.advisory,
        apply
    );
    const securityRuleFilter = useSecurityRuleFilter(
        apply,
        parameters,
        filterRuleValues,
        {
            isDynamic: true,
            dropdownItems: RULE_PRESENCE_OPTIONS.filter(item => item.value !== 'true')
        }
    );

    const activeFilters = {
        filters: buildActiveFilters(parameters, filterRuleValues),
        onDelete: (_, chips, reset) => removeFilters(chips, apply, reset, {}),
        deleteTitle: intl.formatMessage(messages.resetFilters),
        showDeleteButton: !isFilterInDefaultState(parameters, {}, CVE_DETAILS_FILTER_PARAMS)
    };

    const onRowClick = (_key, systemId) => {
        navigate(`/${systemId}?appName=vulnerabilities`);
    };

    return isLoadingInventory ? <Spinner centered /> :
        error?.hasError && !canReadHostsInventory    ? <ErrorHandler code={error?.errorCode} />
            : (<AsynComponent
                appName="inventory"
                ref={inventoryRef}
                module="./ImmutableDevices"
                fallback={<div />}
                store={ReducerRegistry.getStore()}
                onLoad={({ mergeWithEntities }) => {
                    ReducerRegistry.register({
                        ...mergeWithEntities(
                            inventoryEntitiesReducer([]),
                            {
                                page: Number(parameters.page || 1),
                                perPage: DEFAULT_PAGE_SIZE,
                                ...(parameters.sort && {
                                    sortBy: {
                                        key: parameters.sort.replace(/^-/, ''),
                                        direction: parameters.sort.match(/^-/) ? 'desc' : 'asc'
                                    }
                                })
                            }
                        )
                    });
                }}
                key="immutableDevices"
                customFilters={{
                    edgeParams: {
                        ...parameters,
                        host_type: 'edge'
                    }
                }}
                getEntities={getEntities}
                showActions={totalItems !== 0}
                hideFilters={{ all: true, hostGroupFilter: false, operatingSystem: false }}
                noSystemsTable={<EmptyStateNoSystems />}
                mergeAppColumns={mergeAppColumns}
                filterConfig={{
                    items: [
                        searchFilter,
                        securityRuleFilter,
                        advisoryNameFilter
                    ]
                }}
                activeFiltersConfig={activeFilters}
                onRowClick={onRowClick}
            />
            );
};

ImmutableDevices.propTypes = {
    cveName: propTypes.string,
    intl: propTypes.object,
    filterRuleValues: propTypes.object,
    inventoryRef: propTypes.object,
    headerFilters: propTypes.object
};

export default injectIntl(ImmutableDevices);
