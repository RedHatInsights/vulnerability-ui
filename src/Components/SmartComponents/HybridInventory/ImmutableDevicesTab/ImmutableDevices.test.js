import React from 'react';
import '@testing-library/jest-dom';
import { screen, waitFor } from '@testing-library/react';
import { ComponentWithContext } from '../../../../Utilities/TestingUtilities';
import ImmutableDevices from './ImmutableDevices';
import { render } from '@testing-library/react';
import AsynComponent from '@redhat-cloud-services/frontend-components/AsyncComponent';
import { useGetEntities } from './helpers';
import ReducerRegistry from '../../../../Utilities/ReducerRegistry';
import { useRbac } from '../../../../Helpers/Hooks';
import { useSelector } from 'react-redux';

jest.mock('@redhat-cloud-services/frontend-components/AsyncComponent', () => ({
  __esModule: true,
  default: jest.fn((props) => (
    <div {...props} aria-label="immutableDevices-module-mock">
      AsyncComponent
    </div>
  )),
}));

jest.mock('./helpers', () => ({
  ...jest.requireActual('./helpers'),
  useGetEntities: jest.fn(() => {}),
}));

jest.mock('../../../../Helpers/Hooks', () => ({
    ...jest.requireActual('../../../../Helpers/Hooks'),
    useRbac: jest.fn(() => ([ [true], false ] )),
  }));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn((state) => state)
}));

const defaultRenderOptions = { store: ReducerRegistry.getStore() };

const waitAsyncComponent = async () => {
    await waitFor(() => {
        expect(
        screen.getByLabelText('immutableDevices-module-mock')
        ).toBeInTheDocument();
    });
};

afterEach(() => {
    jest.clearAllMocks();
});

describe('ImmutableDevices', () => {
  test('renders without issues', async () => {
   render(
      <ComponentWithContext
        renderOptions={defaultRenderOptions}
      >
        <ImmutableDevices />
      </ComponentWithContext>
  );


    await waitAsyncComponent();
  });

  test('renders with correct custom filters', async () => {
    useSelector.mockReturnValue({                
        page: 1, 
        page_size: 20, 
        sort: "-updated" 
    });

    render(
      <ComponentWithContext
        renderOptions={defaultRenderOptions}
      >
        <ImmutableDevices />
      </ComponentWithContext>
    );

    await waitAsyncComponent();
    expect(AsynComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        customFilters: { 
            edgeParams: { 
                host_type: "edge", 
                page: 1, 
                page_size: 20, 
                sort: "-updated" 
            }
        }
      }),
      {}
    );
  });

  test('renders load ImmutableDevices federated module', async () => {
    render(
      <ComponentWithContext
        renderOptions={defaultRenderOptions}
      >
        <ImmutableDevices />
      </ComponentWithContext>
    );

    await waitAsyncComponent();

    expect(AsynComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        appName: 'inventory',
        module: './ImmutableDevices',
      }),
      {}
    );
  });

  test('should display os and group filter from the ImmutableDevices federated module by default', async () => {
    useSelector.mockReturnValue({ enforceEdgeGroups: false })
    render(
      <ComponentWithContext
        renderOptions={defaultRenderOptions}
      >
        <ImmutableDevices />
      </ComponentWithContext>
    );

    await waitAsyncComponent();

    expect(AsynComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        hideFilters: {
          all: true,
          hostGroupFilter: false,
          operatingSystem: false,
        },
      }),
      {}
    );
  });

  test('should hide group filter from the ImmutableDevices federated module when enforce_edge_group is true', async () => {
    useSelector.mockReturnValue({ enforceEdgeGroups: true, total: 10 })
    render(
      <ComponentWithContext
        renderOptions={defaultRenderOptions}
      >
        <ImmutableDevices />
      </ComponentWithContext>
    );

    await waitAsyncComponent();

    expect(AsynComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        hideFilters: {
          all: true,
          hostGroupFilter: true,
          operatingSystem: false,
        },
      }),
      {}
    );
  });

  test('should call getEntities with correct function arguments', async () => {
    render(
      <ComponentWithContext
        renderOptions={defaultRenderOptions}
      >
        <ImmutableDevices cveName='test-cve'/>
      </ComponentWithContext>
    );

    await waitAsyncComponent();

    expect(useGetEntities).toHaveBeenCalledWith({
        createRows: expect.any(Function),
        id:'test-cve',
    });
  });

  test('should display error state when user is not permitted', async () => {
    useSelector.mockReturnValue({ hasError: true, errorCode: 503 });

    render(
      <ComponentWithContext
        renderOptions={{ store: ReducerRegistry.getStore() }}
      >
        <ImmutableDevices cveName='test-cve'/>
      </ComponentWithContext>
    );

    expect(
        screen.getAllByText('This page is temporarily unavailable')
    ).toHaveLength(1);
  });

  test('should display error state when user is not permitted', () => {
    useRbac.mockReturnValue([[false], false]);
    useSelector.mockReturnValue({});

    render(
      <ComponentWithContext
        renderOptions={{ store: ReducerRegistry.getStore() }}
      >
        <ImmutableDevices />
      </ComponentWithContext>
    );

    expect(
        screen.getByText('You do not have permissions to view or manage Vulnerability')
    ).toBeVisible();
  });
});