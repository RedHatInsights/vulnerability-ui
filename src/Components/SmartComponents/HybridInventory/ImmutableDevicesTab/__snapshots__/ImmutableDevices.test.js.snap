// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImmutableDevices renders without issues 1`] = `
<DocumentFragment>
  <div
    activefiltersconfig="[object Object]"
    appname="inventory"
    aria-label="immutableDevices-module-mock"
    customfilters="[object Object]"
    fallback="[object Object]"
    filterconfig="[object Object]"
    hidefilters="[object Object]"
    module="./ImmutableDevices"
    nosystemstable="[object Object]"
    store="[object Object]"
  >
    AsyncComponent
  </div>
</DocumentFragment>
`;
